query-engine/core/src/telemetry/capturing/capturer.rs

```diff
@@ -0,0 +1,225 @@
+use super::{settings::Settings, storage::Storage};
+use crate::models;
+use opentelemetry::{
+    sdk::{
+        export::trace::SpanData,
+        trace::{Span, SpanProcessor},
+    },
+    trace::{TraceId, TraceResult},
+};
+use std::{collections::HashMap, sync::Arc, sync::Mutex};
+
+/// Capturer determines, based on a set of settings and a trace id, how capturing is going to be handled.
+/// Generally, both the trace id and the settings will be derived from request headers. Thus, a new
+/// value of this enum is created per request.
+#[derive(Debug, Clone)]
+pub enum Capturer {
+    Enabled(Inner),
+    Disabled,
+}
+
+impl Capturer {
+    pub(super) fn new(processor: Processor, trace_id: TraceId, settings: Settings) -> Self {
+        if settings.is_enabled() {
+            return Self::Enabled(Inner {
+                processor,
+                trace_id,
+                settings,
+            });
+        }
+
+        Self::Disabled
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Inner {
+    pub(super) processor: Processor,
+    pub(super) trace_id: TraceId,
+    pub(super) settings: Settings,
+}
+
+impl Inner {
+    pub async fn start_capturing(&self) {
+        self.processor
+            .start_capturing(self.trace_id, self.settings.clone())
+            .await
+    }
+
+    pub async fn fetch_captures(&self) -> Option<Storage> {
+        self.processor.fetch_captures(self.trace_id).await
+    }
+}
+
+/// A [`SpanProcessor`] that captures and stores spans in memory in a synchronized dictionary for
+/// later retrieval
+#[derive(Debug, Clone)]
+pub struct Processor {
+    pub(crate) storage: Arc<Mutex<HashMap<TraceId, Storage>>>,
+}
+
+impl Processor {
+    pub fn new() -> Self {
+        Self {
+            storage: Default::default(),
+        }
+    }
+
+    pub(self) async fn start_capturing(&self, trace_id: TraceId, settings: Settings) {
```

Comment:
I think `pub(self) fn` is equivalent to `fn`, that's the default visibility.

Reply:
noted, thank you. Will remove.

Reply:
Fixed in https://github.com/prisma/prisma-engines/pull/3618/commits/9a545b0a838ce42ddc08107d8e62180492f07888